### PR TITLE
[transactional-test] Record writeset in the golden files.

### PR DESCRIPTION
**Pull request by runtian-zhou**
## Motivation

Record the writesets of execution result in the transactional test. Right now the golden file contains few info in the VM related tests. Adding writesets will allow us to have a strong backward compatibility check as it will force VM to always return the same writeset for existing scripts.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/move-language/move/blob/main/CONTRIBUTING.md#developer-workflow)?

Yes

## Test Plan

`cargo test` and regenerated all golden files

### CODE REVIEW

**Comment by runtian-zhou** [comment]
Not sure if we want to enable this recording only for VM related tests..

File: language/testing-infra/transactional-test-runner/src/vm_test_harness.rs [hunk 1]
```diff
@@ -203,7 +206,11 @@ impl<'a> MoveTestAdapter<'a> for SimpleVMTestAdapter<'a> {
                     format_vm_error(&e)
                 )
             })?;
-        Ok((None, serialized_return_values))
+
+        Ok((
+            Some(format!("ChangeSet: {:?}\n", change_set)),
```
Comment (runtian-zhou) [inline language/testing-infra/transactional-test-runner/src/vm_test_harness.rs:211]: This is the only change. Others are just regenerated golden files

**Review by wrwg — APPROVED** [review]
Great. Can we also cherrypick this into the aptos branch?

**Comment by runtian-zhou** [comment]
> Great. Can we also cherrypick this into the aptos branch?

I think we should. Also would be a great chance to see any slippage happened between main and aptos branch.

**Review by tnowacki — CHANGES_REQUESTED** [review]
Can you put this behind a flag?

File: language/testing-infra/transactional-test-runner/src/vm_test_harness.rs (hunk 1 above)
Comment (tnowacki) [inline language/testing-infra/transactional-test-runner/src/vm_test_harness.rs:211]: We should really do some custom display here, feels a bit weird (and unlike the testing framework) to give just the raw debug dump

**Comment by runtian-zhou** [comment]
Gate the change under a flag and enable it for `move-vm` transactional tests only. With that being said, I would suggest enable the writesets for @tnowacki as they will provide testings over critical backward compatibility guarantee that any network would need.

File: language/testing-infra/transactional-test-runner/src/vm_test_harness.rs [hunk 2]
```diff
@@ -367,6 +385,29 @@ static MOVE_STDLIB_COMPILED: Lazy<Vec<CompiledModule>> = Lazy::new(|| {
...
+        false
+    }
+}
+
 pub fn run_test(path: &Path) -> Result<(), Box<dyn std::error::Error>> {
-    run_test_impl::<SimpleVMTestAdapter>(path, Some(&*PRECOMPILED_MOVE_STDLIB))
+    run_test_impl::<SimpleVMTestAdapter<SkipWriteSetMarker>>(path, Some(&*PRECOMPILED_MOVE_STDLIB))
+}
+
+pub fn run_test_with_writesets(path: &Path) -> Result<(), Box<dyn std::error::Error>> {
```
Comment (runtian-zhou) [inline language/testing-infra/transactional-test-runner/src/vm_test_harness.rs:408]: Add a separate harness to enable recording with writesets.

File: language/testing-infra/transactional-test-runner/src/vm_test_harness.rs (hunk 1 above)
Comment (runtian-zhou) [reply language/testing-infra/transactional-test-runner/src/vm_test_harness.rs]: Any suggestions for what we should dump here?

File: language/testing-infra/transactional-test-runner/src/vm_test_harness.rs [hunk 3]
```diff
@@ -36,13 +37,19 @@ use move_vm_runtime::{
...
 use once_cell::sync::Lazy;
+use std::marker::PhantomData;
 
 const STD_ADDR: AccountAddress = AccountAddress::ONE;
 
-struct SimpleVMTestAdapter<'a> {
+trait RecordWriteSet {
+    fn record_writeset() -> bool;
+}
+
```
Comment (tnowacki) [inline language/testing-infra/transactional-test-runner/src/vm_test_harness.rs:47]: This feels a bit excessive. But I get why you had to do it. Maybe a const generic would work a bit better... but some more thoughts on this below.

**Review by tnowacki — COMMENTED** [review]
I think for like 99% of our tests, these change sets will be empty?

So I would have the following suggestions:
- See how much it would change the tests if we always printed it, but committed any printing when it was empty
- Pretty print the changes for the module/resource sections
- Omit the actual data, and only print the identifier/tag + the op. The data there does not need to be tested here 
- We could maybe print out something nice and slick, like a diff for the account: e.g.
```
0x42: 
  + module A
  + module B
  - resource 0x1::coin::Coin
  + resource 0x1::flag::Flag
  +/- resource 0x3::flargle::Blargle 
```

File: language/testing-infra/transactional-test-runner/src/vm_test_harness.rs [hunk 4]
```diff
@@ -203,7 +213,11 @@ impl<'a> MoveTestAdapter<'a> for SimpleVMTestAdapter<'a> {
                     format_vm_error(&e)
                 )
             })?;
-        Ok((None, serialized_return_values))
+
+        Ok((
+            T::record_writeset().then(|| format!("ChangeSet: {:?}\n", change_set)),
```
[the remark this answers is not in the document]
Comment (jolestar) [reply language/testing-infra/transactional-test-runner/src/vm_test_harness.rs:218]: Maybe use ChangeSet to json? json is more formal than debug output, avoid breaking the testes when debugging output change.

**Comment by wrwg** [comment]
Is this PR still active?

**Comment by runtian-zhou** [comment]
Lemme close this for now just so that we can revisit the golden file situation later.

**Pull request closed without merging**